bundles/org.eclipse.equinox.common/src/org/eclipse/core/runtime/ServiceCaller.java

```diff
@@ -160,6 +166,123 @@ public static <S> boolean callOnce(Class<?> caller, Class<S> serviceType, String
 		}).orElse(Boolean.FALSE);
 	}
 
+	/**
+	 * Not thread save.
+	 * <p>
+	 * Service registrations are obtained upon creation, service instances are get
+	 * on the first request (i.e. the first instance when {@link #first()} or
+	 * {@link #all()} is called. Services that are unregistered in the meantime are
+	 * discarded and newly registered services are ignored.
+	 * </p>
+	 * 
+	 * @param <S> the service type for this caller
+	 * @since 3.17
+	 */
+	public interface ServiceUsageScope<S> extends AutoCloseable { // TODO:? , Supplier<S>
+
+		/**
+		 * Gets the first available service implementation.
+		 */
+		Optional<S> first();
+
+		Stream<S> all();
+
+		@Override
+		void close(); // do not throw (checked) exception
+	}
+
+	/**
+	 * @since 3.17
+	 */
+	public static <T> ServiceUsageScope<T> use(Class<T> service) {
+		Class<?> callerClass = STACK_WALKER.getCallerClass();
+		try {
+			return useService(callerClass, service, null);
+		} catch (InvalidSyntaxException e) {
+			throw new AssertionError("Impossible exception was thrown"); //$NON-NLS-1$
+		}
+	}
+
+	/**
+	 * @since 3.17
+	 */
+	public static <S> ServiceUsageScope<S> use(Class<S> service, String filter) throws InvalidSyntaxException {
+		Class<?> callerClass = STACK_WALKER.getCallerClass();
+		return useService(callerClass, service, filter);
+	}
+
+	private static <S> ServiceUsageScope<S> useService(Class<?> caller, Class<S> serviceType, String filter)
+			throws InvalidSyntaxException {
+		Bundle bundle = FrameworkUtil.getBundle(caller);
+		if (bundle == null) {
+			throw new IllegalStateException(); // TODO: handle differently?
+		}
+		BundleContext ctx = bundle.getBundleContext();
+		if (ctx == null) {
+			throw new IllegalStateException(); // TODO: handle differently?
+		}
+		return use(serviceType, filter, ctx);
+	}
+
+	/**
+	 * @since 3.17
+	 */
+	public static <S> ServiceUsageScope<S> use(Class<S> service, BundleContext context) {
+		try {
+			return use(service, null, context);
+		} catch (InvalidSyntaxException e) {
+			throw new AssertionError("Impossible exception was thrown"); //$NON-NLS-1$
+		}
+	}
+
+	/**
+	 * @since 3.17
+	 */
+	public static <S> ServiceUsageScope<S> use(Class<S> service, String filter, BundleContext context)
+			throws InvalidSyntaxException {
+		ServiceReference<?>[] references = context.getServiceReferences(service.getName(), filter);
```

Comment:
This result in unordered which means the `all` stream will report the ordered characteristic incorrectly and the `first` optional will hold something which is not the "first" as there is no ordering.

So you should sort the references in reverse order so the highest ranked service is first.


```suggestion
		ServiceReference<?>[] references = context.getServiceReferences(service.getName(), filter);
		Arrays.sort(references, Comparator.reverseOrder());
```

This means `first` will always be the highest ranked service and `all` will be ordered so that higher ranked services come before lower ranked services.

You may also want to rename `first` to `highest`.

Reply:
> This means `first` will always be the highest ranked service and `all` will be ordered so that higher ranked services come before lower ranked services.

That's a good suggestion, thank you.

But if I'm not missing anything, for example the result of the `BundleContext.getServiceReference(Class)` does not consider the service ranking does it?

> 
> You may also want to rename `first` to `highest`.

It would indeed by good to reflect that in the name, but I'm not so happy with the name `highest()`. But at the moment I don't have a better one.

Reply:
> But if I'm not missing anything, for example the result of the `BundleContext.getServiceReference(Class)` does not consider the service ranking does it?

It does not. The order of the returned ServiceReferences is unspecified. So you need to sort them.